internal/hud/server/websocket_reader.go

```diff
@@ -0,0 +1,176 @@
+		printer:  printer,
+	}
+}
+func (ls *LogStreamer) Handle(v proto_webview.View) error {
+	fromCheckpoint := logstore.Checkpoint(v.LogList.FromCheckpoint)
+	toCheckpoint := logstore.Checkpoint(v.LogList.ToCheckpoint)
+
+	if fromCheckpoint == -1 {
+		// Server has no new logs to send
+		return nil
+	}
+
+	segments := v.LogList.Segments
+	if fromCheckpoint < ls.checkpoint {
+		// The server is re-sending some logs we already have, so slice them off.
+		deleteCount := ls.checkpoint - fromCheckpoint
+		segments = segments[deleteCount:]
+	}
+
+	// TODO(maia): filter for the resources that we care about (`tilt logs resourceA resourceC`)
+	//   --> and if there's only one resource, don't prefix logs with resource name?
+	for _, seg := range segments {
+		// TODO(maia): secrets???
+		ls.logstore.Append(webview.LogSegmentToEvent(seg, v.LogList.Spans), model.SecretSet{})
+	}
+
+	ls.printer.Print(ls.logstore.ContinuingLines(ls.checkpoint))
+
+	if toCheckpoint > ls.checkpoint {
+		ls.checkpoint = toCheckpoint
+	}
+
+	return nil
+}
+
+func (wsr *WebsocketReader) Listen(ctx context.Context) error {
+	logger.Get(ctx).Debugf("connecting to %s", wsr.url.String())
+
+	c, _, err := websocket.DefaultDialer.Dial(wsr.url.String(), nil)
+	if err != nil {
+		return errors.Wrapf(err, "dialing websocket %s", wsr.url.String())
+	}
+	defer c.Close()
+
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		for {
+			_, data, err := c.ReadMessage()
+			if err != nil {
+				log.Println("🚨 error reading:", err)
+				return
+			}
+
+			v := proto_webview.View{}
+			unmarshaller := jsonpb.Unmarshaler{}
+			err = unmarshaller.Unmarshal(bytes.NewReader(data), &v)
+			if err != nil {
+				log.Println("🚨 error unmarshalling:", err)
+			}
+			err = wsr.handler.Handle(v)
+			if err != nil {
+				log.Println("🚨 handler error:", err)
+			}
+
+			toCheckpoint := v.LogList.ToCheckpoint
+			if toCheckpoint > 0 {
+				// If server is using the incremental logs protocol, ack the
+				// message so the next time the websocket sends data, it only
+				// sends logs from here on forward
+				resp := proto_webview.AckWebsocketRequest{
+					ToCheckpoint:  toCheckpoint,
+					TiltStartTime: v.TiltStartTime,
+				}
+				marshaller := jsonpb.Marshaler{OrigName: false, EmitDefaults: true}
+				respJson, err := marshaller.MarshalToString(&resp)
+				if err != nil {
+					log.Println("🚨 marshalling response:", err)
+				}
+
+				err = c.WriteMessage(websocket.TextMessage, []byte(respJson))
+				if err != nil {
+					log.Println("🚨 sending response:", err)
+				}
+			}
+		}
+	}()
+
+	for {
+		select {
+		case <-done:
+			return nil
+		case <-ctx.Done():
+			err := ctx.Err()
+			if err != context.Canceled {
+				return err
+			}
+
+			// Cleanly close the connection by sending a close message and then
+			// waiting (with timeout) for the server to close the connection.
+			err = c.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""))
```

Comment:
@nicks can you check me here?
1. in general, is this procedure for closing the connection right?
2. should i be writing a `CloseMessage` even if it's not a `context.Canceled` error?

Reply:
i assume you just copied from this?
https://github.com/gorilla/websocket/blob/5ed622c449da6d44c3c8329331ff47a9e5844f71/examples/echo/client.go#L70

but ya, this seems fine to me, but you probably know more about this than i do at this point!